### PR TITLE
[ETFE-3910] Generic Rim rule - Capture response error codes from EIS

### PR DESCRIPTION
**Pull request by james-work-account**

### CODE REVIEW

File: app/controllers/DeclarationController.scala [hunk 1]
```diff
@@ -70,10 +70,19 @@ class DeclarationController @Inject()(
...
       withSubmitCreateMovementModel { submitCreateMovementModel =>
         service.submit(submitCreateMovementModel).flatMap {
-          response =>
+          case Right(response) =>
             logger.debug(s"[onSubmit] response received from downstream service ${response.downstreamService}: ${response.receipt}")
             val updatedAnswers = request.userAnswers.copy(hasBeenSubmitted = true, submittedDraftId = Some(response.submittedDraftId))
             saveAndRedirect(DeclarationPage, LocalDateTime.now(), updatedAnswers, NormalMode)
+          case Left(UnexpectedDownstreamDraftSubmissionResponseError(UNPROCESSABLE_ENTITY)) =>
+            // emcs-tfe returns an UNPROCESSABLE_ENTITY when downstream responds with RIM validation errors,
+            // so we redirect to DraftMovementController to display these errors
```
Comment (LewisMitchell) [inline app/controllers/DeclarationController.scala:79]: Just a general question. Is it possible to get a 422 back and it _not_ be a RIM validation error?

File: app/models/MovementValidationFailure.scala [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+
+package models
+
+import play.api.libs.json.{Format, Json}
+
+case class MovementValidationFailure(
+                                      errorType: Int,
+                                      errorReason: String
```
Comment (LewisMitchell) [inline app/models/MovementValidationFailure.scala:23]: Have we got it confirmed that errorType and errorReason are mandatory within EIS?

File: conf/messages [hunk 3]
```diff
@@ -1096,7 +1096,7 @@ itemDesignationOfOrigin.s200.radio.unprovided.downstream = I don''t want to prov
 itemDesignationOfOrigin.checkYourAnswersLabel = Statement of designation of origin
 itemDesignationOfOrigin.checkYourAnswersLabel.s200 = Statements about the designation of origin and labelling of the product
 itemDesignationOfOrigin.error.geographicalIndication.choice.required = Select a statement about the designation of origin or select ‘I don’t want to provide a statement about the designation of origin’
-itemDesignationOfOrigin.error.spiritMarketedAndLabelled.choice.required = Select the statement about marketing and labelling or select ‘I don’t want to provide a statement about the marketing and labelling of the spirit’
+itemDesignationOfOrigin.error.spiritMarketedAndLabelled.choice.required = Select the statement about marketing and labelling or select ‘I don't want to provide a statement about the marketing and labelling of the spirit’
```
Comment (LewisMitchell) [inline conf/messages:1099]:
```suggestion
itemDesignationOfOrigin.error.spiritMarketedAndLabelled.choice.required = Select the statement about marketing and labelling or select ‘I don’t want to provide a statement about the marketing and labelling of the spirit’
```

File: test/views/DraftMovementViewSpec.scala [hunk 4]
```diff
@@ -60,6 +61,38 @@ class DraftMovementViewSpec extends SpecBase with ViewBehaviours with MovementSu
...
+
+          val validationFailures: Seq[MovementValidationFailure] = Seq(
+            //scalastyle:off magic.number
+            MovementValidationFailure(12, "This is an error."),
+            MovementValidationFailure(13, "This is an error."),
+            MovementValidationFailure(14, "This is an error.")
+            //scalastyle:on magic.number
+          )
+
+          implicit val msgs: Messages = messages(Seq(messagesForLanguage.lang))
```
Comment (LewisMitchell) [inline test/views/DraftMovementViewSpec.scala:74]: Could this implicit be put on line 45? Then it can remove duplication across these 3 tests

**Review by LewisMitchell — CHANGES_REQUESTED** [review]

File: app/controllers/DeclarationController.scala (hunk 1 above)
Comment (james-work-account) [reply app/controllers/DeclarationController.scala:79]: From emcs-tfe, as far as I know no. Play _shouldn't_ throw a 422 without us explicitly throwing it, and I don't think we're explicitly throwing it anywhere at the moment.

File: test/views/DraftMovementViewSpec.scala (hunk 4 above)
Comment (james-work-account) [reply test/views/DraftMovementViewSpec.scala:74]: Sure 🙂

File: app/models/MovementValidationFailure.scala (hunk 2 above)
Comment (LewisMitchell) [reply app/models/MovementValidationFailure.scala:23]: It's a bonus if they are always defined but I suppose if they weren't defined, we wouldn't show it. However, if it were that all errors did not have a reason, we'd end up showing an empty notification banner, unless we guard against that?

I'd say make it optional but only if it does not result in a huge logic change.

File: app/controllers/DeclarationController.scala (hunk 1 above)
Comment (LewisMitchell) [reply app/controllers/DeclarationController.scala:79]: Ah yeah, just checked the spec. I was thinking that EIS _could_ return a 422 despite not being a RIM validation. But that is not the case

**Pull request merged**